// react-native-quick-crypto/lib/module/prime.js
"use strict";

import { NitroModules } from 'react-native-nitro-modules';
import { Buffer } from '@craftzdog/react-native-buffer';
import { binaryLikeToArrayBuffer } from './utils';
let native;
function getNative() {
  if (native == null) {
    native = NitroModules.createHybridObject('Prime');
  }
  return native;
}
function toOptionalArrayBuffer(value) {
  if (value == null) return undefined;
  if (value instanceof ArrayBuffer) return value;
  return binaryLikeToArrayBuffer(value);
}
function bufferToBigInt(buf) {
  let result = 0n;
  for (let i = 0; i < buf.length; i++) {
    result = result << 8n | BigInt(buf[i]);
  }
  return result;
}
function bigIntToBuffer(value) {
  if (value === 0n) return new Uint8Array([0]).buffer;
  const hex = value.toString(16);
  const paddedHex = hex.length % 2 ? '0' + hex : hex;
  const bytes = new Uint8Array(paddedHex.length / 2);
  for (let i = 0; i < bytes.length; i++) {
    bytes[i] = parseInt(paddedHex.substring(i * 2, i * 2 + 2), 16);
  }
  return bytes.buffer;
}
export function generatePrimeSync(size, options) {
  const safe = options?.safe ?? false;
  const add = toOptionalArrayBuffer(options?.add);
  const rem = toOptionalArrayBuffer(options?.rem);
  const result = Buffer.from(getNative().generatePrimeSync(size, safe, add, rem));
  if (options?.bigint) {
    return bufferToBigInt(result);
  }
  return result;
}
export function generatePrime(size, options, callback) {
  if (typeof options === 'function') {
    callback = options;
    options = {};
  }
  const safe = options?.safe ?? false;
  const add = toOptionalArrayBuffer(options?.add);
  const rem = toOptionalArrayBuffer(options?.rem);
  const wantBigint = options?.bigint ?? false;
  getNative().generatePrime(size, safe, add, rem).then(ab => {
    const result = Buffer.from(ab);
    if (wantBigint) {
      callback?.(null, bufferToBigInt(result));
    } else {
      callback?.(null, result);
    }
  }).catch(err => callback?.(err, Buffer.alloc(0)));
}
export function checkPrimeSync(candidate, options) {
  const checks = options?.checks ?? 0;
  const buf = typeof candidate === 'bigint' ? bigIntToBuffer(candidate) : binaryLikeToArrayBuffer(candidate);
  return getNative().checkPrimeSync(buf, checks);
}
export function checkPrime(candidate, options, callback) {
  if (typeof options === 'function') {
    callback = options;
    options = {};
  }
  const checks = options.checks ?? 0;
  const buf = typeof candidate === 'bigint' ? bigIntToBuffer(candidate) : binaryLikeToArrayBuffer(candidate);
  getNative().checkPrime(buf, checks).then(result => callback?.(null, result)).catch(err => callback?.(err, false));
}
