// react-native-quick-crypto/lib/commonjs/certificate.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Certificate = void 0;
var _reactNativeNitroModules = require("react-native-nitro-modules");
var _reactNativeBuffer = require("@craftzdog/react-native-buffer");
var _utils = require("./utils");
let native;
function getNative() {
  if (native == null) {
    native = _reactNativeNitroModules.NitroModules.createHybridObject('Certificate');
  }
  return native;
}
function toArrayBuffer(spkac, encoding) {
  if (typeof spkac === 'string') {
    return (0, _utils.binaryLikeToArrayBuffer)(spkac, encoding || 'utf8');
  }
  return (0, _utils.binaryLikeToArrayBuffer)(spkac);
}
class Certificate {
  static exportChallenge(spkac, encoding) {
    return _reactNativeBuffer.Buffer.from(getNative().exportChallenge(toArrayBuffer(spkac, encoding)));
  }
  static exportPublicKey(spkac, encoding) {
    return _reactNativeBuffer.Buffer.from(getNative().exportPublicKey(toArrayBuffer(spkac, encoding)));
  }
  static verifySpkac(spkac, encoding) {
    return getNative().verifySpkac(toArrayBuffer(spkac, encoding));
  }
}
exports.Certificate = Certificate;
