// react-native-quick-crypto/lib/commonjs/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.defaultCoreCipherList = exports.defaultCipherList = exports.constants = void 0;
const constants = exports.constants = {
  // RSA Padding
  RSA_PKCS1_PADDING: 1,
  RSA_NO_PADDING: 3,
  RSA_PKCS1_OAEP_PADDING: 4,
  RSA_X931_PADDING: 5,
  RSA_PKCS1_PSS_PADDING: 6,
  // RSA PSS Salt Length
  RSA_PSS_SALTLEN_DIGEST: -1,
  RSA_PSS_SALTLEN_MAX_SIGN: -2,
  RSA_PSS_SALTLEN_AUTO: -2,
  // Point Conversion
  POINT_CONVERSION_COMPRESSED: 2,
  POINT_CONVERSION_UNCOMPRESSED: 4,
  POINT_CONVERSION_HYBRID: 6,
  // DH Check
  DH_CHECK_P_NOT_PRIME: 1,
  DH_CHECK_P_NOT_SAFE_PRIME: 2,
  DH_UNABLE_TO_CHECK_GENERATOR: 4,
  DH_NOT_SUITABLE_GENERATOR: 8,
  // OpenSSL Version (3.0.0 = 0x30000000)
  OPENSSL_VERSION_NUMBER: 0x30000000
};
const defaultCoreCipherList = exports.defaultCoreCipherList = 'TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_128_GCM_SHA256';
const defaultCipherList = exports.defaultCipherList = 'TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_128_GCM_SHA256';
